### ISSUE TITLE
Can we merge the cdn storage and pkg/storage interfaces?

### ISSUE THREAD

**Issue by arschles** [gomods/athens#299]
Not sure if this is possible since they serve slightly different use cases (proxy vs. registry), but if we could merge, it could simplify the codebase conceptually.

**Comment by arschles** [comment]
Right now, this is just an item for discussion. Let's hear what people think?

**Comment by robjloranger** [comment]
I can't speak for the other implementations, but for GCP if would mean having two clients for different services. One for object storage and one for database storage. This in itself is not really a big deal, just something to note.

**Comment by marpio** [comment]
The `Deleter` interface is actually same in both and GCP has it already implemented.
I'm  not sure  about the `metadata` part. Do we need to store the URLs? Isn't the `baseURL` provided by each CDN impl. enough?

**Comment by arschles** [comment]
@marpio I think you made a good point there - seems like one of the biggest differences between the two use cases is the getter has to return a redirect URL in the CDN case, but a stream of bytes in the storage/proxy case

**Comment by marpio** [comment]
@arschles yes but I was wondering if we need the `(./pkg/metadata/).Saver`?
Couldn't we just construct the redirect URL for a module?
So i.e. in case of AWS:
`http://d111111abcdef8.cloudfront.net/gomods.io/my/module `
where `gomods.io/my/module` is the module name and `http://d111111abcdef8.cloudfront.net/` is provided by `env.CDNEndpointWithDefault`

**Comment by arschles** [comment]
Check out https://github.com/gomods/athens/issues/325#issuecomment-407580216 for another idea

**Comment by arschles** [comment]
@marpio I don't remember that code super well, but iirc I think we could construct the URL, but `Saver` implementations saves it to the CDN metadata storage too.

**Comment by marpio** [comment]
@arschles I still don't quite understand why do we need the metadata.Saver I'm on phone so not 100% sure but I don't think we use any of the implementations now. Should i.e. the azure storage save the URL in it's Save function? This isn't the case now.
